indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentTransactionalInsertAction.java

```diff
@@ -79,21 +81,32 @@ public static SegmentTransactionalInsertAction appendAction(
       @Nullable DataSourceMetadata endMetadata
   )
   {
-    return new SegmentTransactionalInsertAction(null, segments, startMetadata, endMetadata);
+    return new SegmentTransactionalInsertAction(null, segments, startMetadata, endMetadata, null);
+  }
+
+  public static SegmentTransactionalInsertAction commitMetadataOnlyAction(
+      String dataSource,
+      DataSourceMetadata startMetadata,
+      DataSourceMetadata endMetadata
+  )
+  {
+    return new SegmentTransactionalInsertAction(null, null, startMetadata, endMetadata, dataSource);
   }
 
   @JsonCreator
   private SegmentTransactionalInsertAction(
       @JsonProperty("segmentsToBeOverwritten") @Nullable Set<DataSegment> segmentsToBeOverwritten,
-      @JsonProperty("segments") Set<DataSegment> segments,
+      @JsonProperty("segments") @Nullable Set<DataSegment> segments,
       @JsonProperty("startMetadata") @Nullable DataSourceMetadata startMetadata,
-      @JsonProperty("endMetadata") @Nullable DataSourceMetadata endMetadata
+      @JsonProperty("endMetadata") @Nullable DataSourceMetadata endMetadata,
+      @JsonProperty("dataSource") @Nullable String dataSource
   )
   {
     this.segmentsToBeOverwritten = segmentsToBeOverwritten;
-    this.segments = ImmutableSet.copyOf(segments);
+    this.segments = segments == null ? ImmutableSet.of() : ImmutableSet.copyOf(segments);
```

Comment:
Noting here about rolling upgrade behavior:
- With a new overlord and old task, the old task won't create this action when it created no segments
- With an old overlord and a new task, the new task's metadata commit will fail here when the action is deserialized on the overlord at `ImmutableSet.copyOf(segments)` since `segments` would be null

Reply:
So, we'll need to note that new tasks and old overlords (which is typical during a rolling update) will potentially break if this case is hit? I guess that isn't the end of the world, since it would have broken before anyway. But what should people do about it? Just let it fail repeatedly until the rolling update is done? Should people shut down supervisors before the update and then start them afterwards? What are the odds of this case being hit?

(The above should be answered for the release notes)